firedrake/solving_utils.py

```diff
@@ -214,9 +217,14 @@ def form_function(self, snes, X, F):
 
         assemble(ctx.Fs[lvl], tensor=ctx._Fs[lvl],
                  form_compiler_parameters=problem.form_compiler_parameters)
-        # no mat_type -- it's a vector!
-        for bc in problem.bcs:
-            bc.zero(ctx._Fs[lvl])
+
+        if self.pre_apply_bcs:
```

Comment:
> The `self` in the method signature is misleading. It ought to be `cls`.

That is true as long as the `@classmethod` is there, which means it was misleading before this change. And since it wasn't referenced at all in this function, it could have been a `@staticmethod` instead.